From 23e5ef2978591ba59584d1e34171466039847a13 Mon Sep 17 00:00:00 2001
From: "Ooi, Joyce" <joyce.ooi@intel.com>
Date: Mon, 2 Dec 2019 21:36:05 +0800
Subject: [PATCH 083/120] net: ethernet: altera: fix conversion of scaled_ppm
 to ppb

commit 4013a312932fb2caef5737ad5e569864a3c24a2b from
https://github.com/altera-opensource/linux-socfpga.git

This patch fixes the conversion of scaled_ppm to ppb.

Signed-off-by: Ooi, Joyce <joyce.ooi@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/net/ethernet/altera/intel_fpga_tod.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ethernet/altera/intel_fpga_tod.c b/drivers/net/ethernet/altera/intel_fpga_tod.c
index 3771597642da..e3af41077057 100644
--- a/drivers/net/ethernet/altera/intel_fpga_tod.c
+++ b/drivers/net/ethernet/altera/intel_fpga_tod.c
@@ -98,7 +98,8 @@ static int intel_fpga_tod_adjust_fine(struct ptp_clock_info *ptp,
 	unsigned long flags;
 	unsigned long rate;
 	int ret = 0;
-	u64 ppb;
+	s64 ppb;
+	u64 new_ppb;
 
 	rate = clk_get_rate(priv->tod_clk);
 
@@ -107,9 +108,9 @@ static int intel_fpga_tod_adjust_fine(struct ptp_clock_info *ptp,
 	ppb *= 125;
 	ppb >>= 13;
 
-	ppb += NOMINAL_PPB;
+	new_ppb = (s32)ppb + NOMINAL_PPB;
 
-	tod_period = div_u64_rem(ppb << 16, rate, &tod_rem);
+	tod_period = div_u64_rem(new_ppb << 16, rate, &tod_rem);
 	if (tod_period > TOD_PERIOD_MAX) {
 		ret = -ERANGE;
 		goto out;
-- 
2.31.1

